core/src/main/java/macrobase/analysis/sample/AChao.java

```diff
@@ -24,17 +51,56 @@ public AChao(int capacity, Random random) {
         this.random = random;
     }
 
+    private void updateOverweightItems() {
+        while(!overweightItems.isEmpty()) {
+            OverweightItem<T> ow = overweightItems.peek();
+            if(reservoirCapacity * ow.weight / runningCount <= 1) {
+                overweightItems.poll();
+                insert(ow.item, ow.weight);
+            } else {
+                break;
+            }
+        }
+    }
+
     public final List<T> getReservoir() {
+        updateOverweightItems();
+
+        if(!overweightItems.isEmpty()) {
+            // overweight items always make it in the sample
+            List<T> ret = overweightItems.stream().map(i -> i.item).collect(Collectors.toList());
+
+            assert (ret.size() <= reservoirCapacity);
```

Comment:
Ah, sorry, yeah, this should be better documented. (not an excuse, but it is also really hard to grok in the paper).

`updateOverweightItems` has the logic that will do the pruning.

I missed the above stat test etc., but yeah, we should do that :)